src/modm/platform/adc/samg/adc.hpp.in

```diff
@@ -81,33 +89,59 @@ class Adc : modm::Adc
 			frequency,                              // desired adc frequency
 			0,                                      // lowest prescaler value
 			255,                                    // highest prescaler value
+%% if target.family == "e7x/s7x/v7x"
+			[](uint32_t x) { return (x + 1); }      // transform function
+%% else
 			[](uint32_t x) { return 2 * (x + 1); }  // transform function
+%% endif
 		);
 		static_assert(result.frequency <= MaxAdcFrequency,
 					  "Generated ADC frequency is above maximum frequency!");
 		assertBaudrateInTolerance<result.frequency, frequency, tolerance>();
 
-		ClockGen::enable<ClockPeripheral::Adc>();
-
-		Regs()->ADC_MR = ADC_MR_PRESCAL(result.index) | ADC_MR_TRANSFER(2);
+		ClockGen::enable<ClockPeripheral::{{type | capitalize}}{{instance}}>();
+
+		Regs()->{{reg}}_MR =
+			{{reg}}_MR_PRESCAL(result.index) |
+%% if target.family == "e7x/s7x/v7x"
+			// values from vendor hal code and application note
+			// {{reg}}_MR_TRANSFER(1) is a forbidden value according to the datasheet
+			{{reg}}_MR_TRANSFER(1) |
```

Comment:
But but it's forbidden???2?

Reply:
I have completely stopped trusting the correctness of the ADC section in the datasheet. It states for the `TRACKTIM` field:

"Inherent tracking time is always 15 AFE clock cycles. Do not modify this field."

If I follow their guidance and leave `TRACKTIM` on the reset value of 0, the ADC will read values close to half scale about 50% of the time when its input is connected to GND.

The examples in the app note number 44046 ("Using the Analog Front End in the SAM V7/E7/S7
MCUs") use `AFEC_MR_TRACKTIM(2)` which according to the text means `TRACKTIM + 1 = 3` AFE clock cycles which is not at all equal to 15. That example also has the illegal value of `AFEC_MR_TRANSFER(1)`. Those values are also what the ASF3 vendor HAL has as defaults.

Assuming the app note is right about the meaning of the `TRACKTIM` field it has to be set to 14 to achieve the recommended tracking time of 15 AFE clock cycles.

There are actually 3 different vendor HAL options offered right now, ASF3, ASF4 and MPLAB Harmony.
ASF4 can only be obtained by registering and downloading a Windows IDE, so I skipped it. The others can be found somewhere on github.

The MPLAB HAL has `AFEC_MR_TRANSFER(1)` and `AFEC_MR_TRACKTIM(15U)` which would equate to 16 cycles if the app note was right.

This is all a bit too confusing. I'll go with `AFEC_MR_TRACKTIM(14)` and the datasheet recommended value of `AFEC_MR_TRANSFER(2)` until proven otherwise. At least the ADC seems to do the right thing.